src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+	} else {
+		/* Construct NBD_OPT_EXPORT_NAME option */
+		opt->opt_export_name.request.request_magic =
+			htonll ( NBD_OPT_REQ_MAGIC );
+		opt->opt_export_name.request.option =
+			htonl ( NBD_OPT_EXPORT_NAME );
+		opt->opt_export_name.request.length = htonl ( name_len );
+		strcpy ( opt->opt_export_name.export_name, nbd->export_name );
+	}
+
+	DBGCIO ( nbd, "transmit:\n" );
+	DBGCIO_HD ( nbd, opt, opt_len );
+
+	/* Deliver packet */
+	if ( ( rc = xfer_deliver_iob ( &nbd->socket,
+				       iob_disown ( iobuf ) ) ) != 0 ) {
+		DBGC ( nbd, "NBD %p cannot transmit: %s\n",
+		       nbd, strerror ( rc ) );
+		return rc;
+	}
+
+	nbd->rx_state = nbd->use_opt_go ?
+		NBD_RX_NEG_OPT_INFO : NBD_RX_NEG_EXP_NAME;
+	nbd->tx_state = NBD_TX_IDLE;
+	return 0;
+}
+
+/**
+ * Process the block_read_capacity command
+ *
+ * @v nbd		NBD session
+ */
+void nbd_tx_cmd_brc ( struct nbd_session *nbd ) {
+	struct nbd_command *nbdcmd = nbd->command;
+	struct block_device_capacity capacity;
+
+	/* Sanity check */
+	assert ( nbdcmd != NULL && nbdcmd->type == NBD_BLOCK_READ_CAPACITY );
+
+	capacity.blocks = nbd->export_size / NBD_BLOCK_SIZE;
+	capacity.blksize = NBD_BLOCK_SIZE;
+	/* Use reasonable amount of data */
+	capacity.max_count = 16 * 1024 / NBD_BLOCK_SIZE;
+
+	/* Report block device capacity */
+	block_capacity ( &nbdcmd->block, &capacity );
+
+	nbdcmd_close (nbdcmd, 0);
+	nbd->tx_state = NBD_TX_IDLE;
+}
+
+/**
+ * Transmit a command header
+ *
+ * @v nbd		NBD session
+ * @ret rc		Return status code
+ */
+int nbd_tx_cmd_header ( struct nbd_session *nbd ) {
+	struct nbd_command *nbdcmd = nbd->command;
+	struct nbd_proto_trans_request request;
+	int rc;
+
+	/* Sanity check */
+	assert ( nbdcmd != NULL &&
+		 ( nbdcmd->type == NBD_BLOCK_READ ||
+		   nbdcmd->type == NBD_BLOCK_WRITE ) );
+
+	request.request_magic = htonl ( NBD_REQUEST_MAGIC );
+	request.flags = 0;
+	if ( nbdcmd->type == NBD_BLOCK_READ )
+		request.type = htons ( NBD_CMD_READ );
+	else
+		request.type = htons ( NBD_CMD_WRITE );
+	request.handle = (uint64_t) (intptr_t) nbdcmd;
+	request.offset = htonll ( nbdcmd->lba * NBD_BLOCK_SIZE );
+	request.length = htonl ( nbdcmd->count * NBD_BLOCK_SIZE );
+
+	/* Deliver request */
+	if ( ( rc = xfer_deliver_raw ( &nbd->socket, &request,
+				       sizeof ( request ) ) ) != 0 ) {
+		DBGC ( nbd, "NBD %p cannot transmit: %s\n",
+		       nbd, strerror ( rc ) );
+		return rc;
+	}
+
+	if ( nbdcmd->type == NBD_BLOCK_READ )
+		nbd->tx_state = NBD_TX_IDLE;
+	else
+		nbd->tx_state = NBD_TX_CMD_DATA;
+	return 0;
+}
+
+/**
+ * Transmit a command data
+ *
+ * @v nbd		NBD session
+ * @ret rc		Return status code
+ */
+int nbd_tx_cmd_data ( struct nbd_session *nbd ) {
+	struct nbd_command *nbdcmd = nbd->command;
+	size_t len;
+	struct io_buffer *iobuf;
+	int rc;
+
+	/* Sanity check */
+	assert ( nbdcmd != NULL && nbdcmd->type == NBD_BLOCK_WRITE );
+
+	len = nbdcmd->data_len - nbdcmd->data_offset;
+	/* Always send 512-byte data chunks */
+	if ( len > 512 )
+		len = 512;
```

Comment:
Use a named constant.

Reply:
iSCSI code uses the same unnamed constant...
OK, fixed